### ISSUE TITLE
Forever initializing after 1.8

### ISSUE THREAD

**Issue by frankfrancois** [bang-olufsen/bang_olufsen-hacs#16]
Hi

After upgrade to 1.8 my theatre keeps initializing forever. Attached screenshot and log file with debug enabled. The log file does not give me any clue to what is causing the problem. I've tried to delete the theatre device and added it again. Can someone point me to a direction where I can see the rootcause to the init problem.

Kind regards
Frank

Attached files:
![init](https://github.com/bang-olufsen/bangolufsen-hacs/assets/151915865/51aed2fc-d54a-49c2-91dc-bbbe141b221c)
[home-assistant_2024-01-26T17-16-21.144Z.log](https://github.com/bang-olufsen/bangolufsen-hacs/files/14067619/home-assistant_2024-01-26T17-16-21.144Z.log)

The hiccup:

2024-01-26 18:05:00.471 ERROR (MainThread) [homeassistant.config_entries] Error setting up entry Beosound Theatre-35857485 for bangolufsen
Traceback (most recent call last):
File "/usr/src/homeassistant/homeassistant/config_entries.py", line 406, in async_setup
result = await component.async_setup_entry(hass, self)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/config/custom_components/bangolufsen/__init__.py", line 93, in async_setup_entry
if not await client.check_websocket_connection():
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/mozart_api/mozart_client.py", line 200, in check_websocket_connection
async with session.ws_connect(f"ws://{self.host}:9339/") as websocket:
File "/usr/local/lib/python3.11/site-packages/aiohttp/client.py", line 1187, in __aenter__
self._resp = await self._coro
^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/aiohttp/client.py", line 825, in _ws_connect
resp = await self.request(
^^^^^^^^^^^^^^^^^^^
File "/usr/local/lib/python3.11/site-packages/aiohttp/client.py", line 601, in _request
await resp.start(conn)
File "/usr/local/lib/python3.11/site-packages/aiohttp/client_reqrep.py", line 960, in start
with self._timer:
File "/usr/local/lib/python3.11/site-packages/aiohttp/helpers.py", line 735, in __exit__
raise asyncio.TimeoutError from None
TimeoutError
2024-01-26 18:05:00.521 ERROR (MainThread) [homeassistant] Error doing job: Unclosed client session

**Comment by mj23000** [comment]
Sorry to hear that you're having problems.

I'm not able to replicate this behavior on a Balance,  Emerge, Level or a Theatre. 

What version of Mozart is the Theatre running? And I'm guessing it was working in 1.6.2? Was the Theatre added to Home Assistant using Discovery or by manually entering the IP address?

It looks like Home Assistant can't establish a WebSocket connection to the Theatre. 
A difference between 1.6.2 and 1.8.0, is that 1.8.0 checks if a WebSocket connection can be made before starting the integration, which is where your setup is failing. The code to check connection and actually establish connection is not very different, so I wouldn't expect there to be any problems. 

Kind regards

**Comment by FrankLundgreen** [comment]
My Theatre is running software version 3.2.3.2. I tried to delete all entities, deleted the HACS integration, restartet my Synology (which runs HA in docker), added the 1.8.1 and the problem is the same - regardless of adding it in discovery mode or manually by the static IP. My beolab 28 and Level is added ok in a few secs.

Something definitely changed in 1.8.x. When I downgrade to 1.6.2 it adds my Theatre in a few secs. By the way - I am a programmer (java). If I can help troubleshoot I would be happy to help (the reward would be that I am not stuck on 1.6.x 🤗 )

Here's my Docker HA stack + versions:

![docker](https://github.com/bang-olufsen/bangolufsen-hacs/assets/151915865/72242f60-b92f-4abe-82e4-03d290297097)

Kind Regards
Frank

**Comment by mj23000** [comment]
So, the API actually got a big change from 1.6 to 1.8. Instead of using Threads for the WebSocket listeners, we are now using Asyncio Tasks and also a new way of handling notifications. But your setup is failing before this is even relevant. 

Can I get you to comment out line 93-96:
```python
if not await client.check_websocket_connection():
    raise ConfigEntryNotReady(
        f"Unable to connect to {entry.title} WebSocket notification channel"
    )
```
in \_\_init__.py in version 1.8, reboot Home Assistant and see if the Theatre works as expected.

Kind regards

**Comment by FrankLundgreen** [comment]
Yeah Markus. That worked.

By the way. Last night all speakers where updated to 3.4.1.4. At first I just installed 1.8.1, without the code change, just to see if it worked. For some crazy reason the Theatre was added ok, but then it was my Beolab which got stuck in initializing. I commented out the code in init_py and all of my use cases now works on 1.8.1:

* Press on a FUGA light switch turns on Live TV and lights in my living room
* Press on favourite buttons on my Level groups other speakers in the house
* Press on light control on my Beoremote One controls the lights in my living room 
* Volume control on all speakers from my HA app or FUGA switch

Kind Regards
Frank

**Comment by mj23000** [comment]
Good to hear!

Looks like we've found the culprit (WebSocket connection can't be established outside of a Task/Thread sometimes?). 
I've reworked the way API and WebSocket connection is checked. Could I get you to test this?

In \_\_init__.py:

Replace `async_setup_entry`
With this:
```python
async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up from a config entry."""
    # Remove casts to str
    assert entry.unique_id

    # Create device now as BangOlufsenWebsocket needs a device for debug logging, firing events etc.
    # And in order to ensure entity platforms (button, binary_sensor) have device name before the primary (media_player) is initialized
    device_registry = dr.async_get(hass)
    device_registry.async_get_or_create(
        config_entry_id=entry.entry_id,
        identifiers={(DOMAIN, entry.unique_id)},
        name=entry.title,
        model=entry.data[CONF_MODEL],
    )
    client = MozartClient(host=entry.data[CONF_HOST], websocket_reconnect=True)

    # Check API and WebSocket connection
    if not await client.check_device_connection():
        await client.close_api_client()
        raise ConfigEntryNotReady(f"Unable to connect to {entry.title}")

    # Initialize coordinator
    coordinator = BangOlufsenCoordinator(hass, entry, client)
    await coordinator.async_config_entry_first_refresh()

    # Add the websocket and API client
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = BangOlufsenData(
        coordinator, client
    )

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    # Start WebSocket connection when all entities have been initialized
    entry.async_create_background_task(
        hass,
        _start_websocket_listener(hass, entry, hass.data[DOMAIN][entry.entry_id]),
        f"{DOMAIN}-{entry.unique_id}-websocket_starter",
    )

    return True
```

In manifest.json:
Replace the "requirements" key's value with "mozart_api-3.2.1.150.7.tar.gz".
```json
"requirements": ["mozart_api-3.2.1.150.7.tar.gz"],
```

In config/:

Place the [mozart_api-3.2.1.150.7.tar.gz](https://github.com/bang-olufsen/bangolufsen-hacs/files/14113120/mozart_api-3.2.1.150.7.tar.gz) file in the `config` folder.
Like so:
<img width="363" alt="image" src="https://github.com/bang-olufsen/bangolufsen-hacs/assets/24912668/3cf53ca8-56d9-4cc6-a857-fe9c3fd88ea4">

**Comment by FrankLundgreen** [comment]
I've tested that. Attached edited files + logfile with debug.
/frank

[manifest.json](https://github.com/bang-olufsen/bangolufsen-hacs/files/14113783/manifest.json)
[__init__.py.txt](https://github.com/bang-olufsen/bangolufsen-hacs/files/14113802/__init__.py.txt)
[home-assistant.log](https://github.com/bang-olufsen/bangolufsen-hacs/files/14113804/home-assistant.log)

![config](https://github.com/bang-olufsen/bangolufsen-hacs/assets/151915865/003494ee-e8ef-4699-bfb8-9d5f8f4c412a)

![init](https://github.com/bang-olufsen/bangolufsen-hacs/assets/151915865/1ea02adc-9b33-482c-b6f2-ad8709023bea)

**Comment by mj23000** [comment]
Thank you for your patience. It looks like it may be the WebSocket library that I am using. 
I've replaced the WebSocket library. Can I get you to do the following?

In \_\_init__.py

Replace `_start_websocket_listener` and `async_setup_entry` with:

```python
async def _start_websocket_listener(
    hass: HomeAssistant, entry: ConfigEntry, data: BangOlufsenData
) -> None:
    """Start WebSocket listener when all entities have been initialized."""
    entity_registry = er.async_get(hass)

    while True:
        # Get all entries for entities and filter all disabled entities
        entries = er.async_entries_for_config_entry(entity_registry, entry.entry_id)
        expected_entities = len(
            [entry for entry in entries if not entry.disabled or not entry.disabled_by]
        )

        # Check if all entities and platforms have been initialized and start WebSocket listener
        if (
            expected_entities == data.entities_initialized
            and len(PLATFORMS) == data.platforms_initialized
        ):
            await data.client.connect_notifications(remote_control=True, reconnect=True)
            return

        await asyncio.sleep(0)


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up from a config entry."""
    # Remove casts to str
    assert entry.unique_id

    # Create device now as BangOlufsenWebsocket needs a device for debug logging, firing events etc.
    # And in order to ensure entity platforms (button, binary_sensor) have device name before the primary (media_player) is initialized
    device_registry = dr.async_get(hass)
    device_registry.async_get_or_create(
        config_entry_id=entry.entry_id,
        identifiers={(DOMAIN, entry.unique_id)},
        name=entry.title,
        model=entry.data[CONF_MODEL],
    )

    client = MozartClient(host=entry.data[CONF_HOST])

    # Check API and WebSocket connection
    if not await client.check_device_connection():
        await client.close_api_client()
        raise ConfigEntryNotReady(f"Unable to connect to {entry.title}")

    # Initialize coordinator
    coordinator = BangOlufsenCoordinator(hass, entry, client)
    await coordinator.async_config_entry_first_refresh()

    # Add the coordinator and API client
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = BangOlufsenData(
        coordinator, client
    )

    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)

    # Start WebSocket connection when all entities have been initialized
    entry.async_create_background_task(
        hass,
        _start_websocket_listener(hass, entry, hass.data[DOMAIN][entry.entry_id]),
        f"{DOMAIN}-{entry.unique_id}-websocket_starter",
    )

    return True
```
And in manifest.json

```json
"requirements": ["mozart_api-3.2.1.150.8.tar.gz"],
```

And place [mozart_api-3.2.1.150.8.tar.gz](https://github.com/bang-olufsen/bangolufsen-hacs/files/14127544/mozart_api-3.2.1.150.8.tar.gz) in the config/ folder

**Comment by FrankLundgreen** [comment]
Good evening 🤗 

After editing the files + restart HA the first thing I noticed is that HA only discover my Beolab (1) and icon in (2) has gone. It does not discover the Level and Theatre any longer:

![overview](https://github.com/bang-olufsen/bangolufsen-hacs/assets/151915865/04f44cc4-8298-47a7-b2b3-c628705873aa)

I added Level+Theatre manually by their static IP. When I entered the Beolab 28 IP it finds it in a few secs. But before I got to place in in LivingRoom (i.e Stue) it failed in the background while initializing. You can see it here;

![setup](https://github.com/bang-olufsen/bangolufsen-hacs/assets/151915865/008bcf93-d5ff-42f1-99c8-0366dde6eb51)

There is no mention of Beolab in the log?
[home-assistant.log](https://github.com/bang-olufsen/bangolufsen-hacs/files/14129262/home-assistant.log)

By the way. I'm thinking about adding a Beoremote Halo in my setup. Will I be able to listen to events from it in HA - just like I do on my Beoremote One (lys, styring)

More debugging? Just write again. I'm happy to help.

**Comment by mj23000** [comment]
Have you updated to the latest Home Assistant Core beta?

The missing icon is expected as a simplified version of this integration has been merged into the official release, but it has been renamed to bang_olufsen, making the icons for the current name "bangolufsen" disappear. This will be fixed in a 2.0.0 release soon.

The Beoremote Halo is currently not directly supported, but can be made to work [see video](https://www.youtube.com/watch?v=jsh1S1GP0HI). There is some help to find in the Beotalk discord channel.

I'l try to find a pair of Beolab 28 to test on.

**Comment by mj23000** [comment]
I'll have to get an overview of your setup. I'm guessing you have a pair of Beolab 28 that are stereo paired. Do you have your Beolab 28 connected to the Beosound Theatre with Power Link? And is it wireless or wired?

**Comment by FrankLundgreen** [comment]
My Beolab 28's are stereo paired and connected to my Theatre with wireless powerlink 🤗

**Comment by frankfrancois** [comment]
Just to follow up. Status right now, on 2.0.0, is that everything works except my Beolab 28's which I cannot add. I tried to trace it locally. Replaced the striped down version in Home Assistant Core 2024.2 with 2.0.0 and it is obvious that the hiccup is a failing websocket connection. My trace hangs here:

![trace](https://github.com/bang-olufsen/bang_olufsen-hacs/assets/151915865/e3f2dc9d-635d-4c47-9b17-857f09ffc656)

And Postman confirms that the ws connection is not possible:

![postman](https://github.com/bang-olufsen/bang_olufsen-hacs/assets/151915865/6d119b68-e4ef-45ab-9988-3f90e62f0c94)

It is over my paygrade to figure out what is wrong 🤗 I have no idea if it is the wireless powerlink that causes this. I reckon that a setup without powerlink would be more straight forward.

**Comment by mj23000** [comment]
At this point it looks like it is a setup issue. I've Tried to replicate the Theatre and wireless powerlink setup with the stereo paired Beolab 28s and it was working as expected. Beolink with a Level also worked with this setup.

What I would recommed is a factory reset of both of the Beolab 28 and if that doesn't fix the problem, then also the Theatre.

**Comment by frankfrancois** [comment]
I restored all products back to the official releases (and made a opt-out of the beta program). That fixed the problem. Now all products are added to HA (and I can see that it can now make a websocket connection to my beolab). I learned a lot during debugging. Thanks a lot for your support 🤗
